From 61f271dc4f13b4a883dced95a405239962893732 Mon Sep 17 00:00:00 2001
From: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
Date: Tue, 7 Jan 2025 14:50:51 -0800
Subject: [PATCH 02/16] PENDING: net: PCS: Add 10GBaseR support in XPCS driver

Qcom supports only USXGMII in the current implementation.
Added support to disable Cl37 PCS autoneg and enable 10GbaseR
for Fixed link cases.

Change-Id: I0ed4b50ba76ab09461ae09c4f9b4b873595776f4
Signed-off-by: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
(cherry picked from commit 984c94770cf5d2095c15042c67c2ad54dbcb915e)
---
 drivers/net/pcs/pcs-xpcs-qcom.c | 30 +++++++++++++++++++++++++++---
 include/linux/pcs-xpcs-qcom.h   |  3 ++-
 2 files changed, 29 insertions(+), 4 deletions(-)

diff --git a/drivers/net/pcs/pcs-xpcs-qcom.c b/drivers/net/pcs/pcs-xpcs-qcom.c
index 159e01348237..a908c928b7e7 100644
--- a/drivers/net/pcs/pcs-xpcs-qcom.c
+++ b/drivers/net/pcs/pcs-xpcs-qcom.c
@@ -2,7 +2,7 @@
 /*
  * Synopsys DesignWare XPCS platform device driver
  *
- * Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
+ * Copyright (c) 2024-2025 Qualcomm Innovation Center, Inc. All rights reserved.
  */
 
 #include <linux/delay.h>
@@ -44,12 +44,27 @@ static const int xpcs_usxgmii_features[] = {
 	__ETHTOOL_LINK_MODE_MASK_NBITS,
 };
 
+static const int xpcs_10gbaser_features[] = {
+	ETHTOOL_LINK_MODE_Pause_BIT,
+	ETHTOOL_LINK_MODE_Asym_Pause_BIT,
+	ETHTOOL_LINK_MODE_10000baseSR_Full_BIT,
+	ETHTOOL_LINK_MODE_10000baseLR_Full_BIT,
+	ETHTOOL_LINK_MODE_10000baseLRM_Full_BIT,
+	ETHTOOL_LINK_MODE_10000baseER_Full_BIT,
+	__ETHTOOL_LINK_MODE_MASK_NBITS,
+};
+
 static const phy_interface_t xpcs_usxgmii_interfaces[] = {
 	PHY_INTERFACE_MODE_USXGMII,
 };
 
+static const phy_interface_t xpcs_10gbaser_interfaces[] = {
+	PHY_INTERFACE_MODE_10GBASER,
+};
+
 enum {
 	DW_XPCS_USXGMII,
+	DW_XPCS_10GBASER,
 	DW_XPCS_INTERFACE_MAX,
 };
 
@@ -355,6 +370,8 @@ static int qcom_xpcs_do_config(struct dw_xpcs_qcom *qxpcs, phy_interface_t inter
 		if (ret < 0)
 			return ret;
 		break;
+	case DW_10GBASER:
+		break;
 	default:
 		XPCSERR("Incompatible Autonegotiation mode\n");
 		return -EINVAL;
@@ -468,7 +485,7 @@ static int qcom_xpcs_select_mode(struct dw_xpcs_qcom *qxpcs, phy_interface_t int
 
 	g_interface = interface;
 
-	if (interface == PHY_INTERFACE_MODE_USXGMII) {
+	if (interface == PHY_INTERFACE_MODE_USXGMII || interface == PHY_INTERFACE_MODE_10GBASER) {
 		ret = qcom_xpcs_read(qxpcs, DW_SR_MII_PCS_CTRL2);
 		if (ret < 0)
 			goto out;
@@ -495,7 +512,8 @@ static int qcom_xpcs_select_mode(struct dw_xpcs_qcom *qxpcs, phy_interface_t int
 		if (ret < 0)
 			goto out;
 
-		qcom_xpcs_write(qxpcs, DW_VR_MII_PCS_DIG_CTRL1, ret | DW_USXGMII_EN);
+		if (interface == PHY_INTERFACE_MODE_USXGMII)
+			qcom_xpcs_write(qxpcs, DW_VR_MII_PCS_DIG_CTRL1, ret | DW_USXGMII_EN);
 
 		return 0;
 	}
@@ -563,6 +581,12 @@ static const struct xpcs_compat synopsys_xpcs_compat[DW_XPCS_INTERFACE_MAX] = {
 		.num_interfaces = ARRAY_SIZE(xpcs_usxgmii_interfaces),
 		.an_mode = DW_AN_C37_USXGMII,
 	},
+	[DW_XPCS_10GBASER] = {
+		.supported = xpcs_10gbaser_features,
+		.interface = xpcs_10gbaser_interfaces,
+		.num_interfaces = ARRAY_SIZE(xpcs_10gbaser_interfaces),
+		.an_mode = DW_10GBASER,
+	},
 };
 
 static const struct xpcs_id xpcs_id_list[] = {
diff --git a/include/linux/pcs-xpcs-qcom.h b/include/linux/pcs-xpcs-qcom.h
index d9eb927d70be..9d4cb034f873 100644
--- a/include/linux/pcs-xpcs-qcom.h
+++ b/include/linux/pcs-xpcs-qcom.h
@@ -2,7 +2,7 @@
 /*
  * Synopsys DesignWare XPCS platform device driver
  *
- * Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
+ * Copyright (c) 2024-2025 Qualcomm Innovation Center, Inc. All rights reserved.
  */
 
 #ifndef __LINUX_PCS_XPCS_QCOM_H
@@ -13,6 +13,7 @@
 
 /* AN mode */
 #define DW_AN_C37_USXGMII		1
+#define DW_10GBASER			5
 
 struct xpcs_id;
 
